### PR TITLE
Better environment initialisation

### PR DESCRIPTION
**Pull request by aaronpowell**
Making it so when you set the venv it will initialise that as well, assuming it doesn't exist (opt-out available via optional arg).

~~When the builder is building the `IPythonEnvironment` it will now try to install packages with pip, assuming a `requirements.txt` is found.~~ Rolled back this change and will tackle it in a new branch.

Some refactoring also.

- **Moving stuff around to be more C#-y**
- **Initializing the venv and providing a way to opt-out of doing so**
- ~~**Using pip to install deps if requirements.txt is found**~~
- **Moving the version of packages to props file for easier local dev**
- **Adding NuGet version to props file and enabling SourceLink for better debugging**
- **I think that adds the missing dependency**

### CODE REVIEW

File: src/CSnakes.Runtime/CSnakes.Runtime.csproj [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
     <ItemGroup>
         <PackageReference Include="pythonnet" Version="3.1.0-preview2024-07-03" />
+        <PackageReference Include="Microsoft.SourceLink.GitHub" Version="8.0.0" PrivateAssets="All"/>
```
Comment (tonybaloney) [inline src/CSnakes.Runtime/CSnakes.Runtime.csproj:9]: What does this do?

File: src/CSnakes.Runtime/PythonEnvironmentBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        {
+            WorkingDirectory = home,
+            FileName = "pip",
+            Arguments = "install -r requirements.txt"
+        };
+
+        if (virtualEnvironmentLocation is not null)
+        {
+            string venvScriptPath = Path.Combine(virtualEnvironmentLocation, "Scripts");
+            startInfo.FileName = Path.Combine(venvScriptPath, "pip.exe");
```
Comment (tonybaloney) [inline src/CSnakes.Runtime/PythonEnvironmentBuilder.cs:78]: This is Windows specific

Comment (aaronpowell) [reply src/CSnakes.Runtime/PythonEnvironmentBuilder.cs:78]: ah good point, it shouldn't be `.exe` off Windows. Is the rest of the path more consistent?

Comment (tonybaloney) [reply src/CSnakes.Runtime/PythonEnvironmentBuilder.cs:78]: It's bin/pip for Nix and Scripts\pip.exe for Windows. But see latter comment about SoC

**Comment by tonybaloney** [comment]
> > This is much cleaner, I wonder if we should move the ensure virtual environment code out though. It's going to be more complicated than pip install for a lot of setups, also for AI/ML projects Conda is likely to be the package manager just as much as pop.
> 
> 
> 
> Sure, we could have something that allows you to specify the package manager rather than it assuming pip. Or we could go down the path of making it more DI based, so you register a package manager like:
> 
> 
> 
> ```csharp
> 
> var builder = PythonEnvironmentBuilder.FromNuGet("3.12.3");
> 
> 
> 
> builder.Services.AddPip();
> 
> 
> 
> // and so on
> 
> ```
> 
> 

I think we'd inevitably want it to be DI/pluggable, if we want to implement this at all. Python devs are used to having to setup the virtual environment as a step before the code will work and the steps for doing so will vary between projects. 
I think that taking it out of this PR for now makes sense since it contains lots of other important changes then we can discuss the design of a package initialization tool.

File: src/CSnakes.Runtime/CSnakes.Runtime.csproj (hunk 1 above)
Comment (aaronpowell) [reply src/CSnakes.Runtime/CSnakes.Runtime.csproj:9]: https://learn.microsoft.com/dotnet/standard/library-guidance/sourcelink

Makes debugging into external packages easier.

File: src/CSnakes.Runtime/PythonEnvironmentBuilder.cs (hunk 2 above)
Comment (aaronpowell) [reply src/CSnakes.Runtime/PythonEnvironmentBuilder.cs:78]: Ah right. I'll pull this part out as a different PR.

**Comment by aaronpowell** [comment]
Having to provision the venv makes sense at least, and it simplifies the CI environment.

I also think it'll be good to have package install as part of the builder pipeline, as it means you don't have to run additional stuff outside yourself before the .NET process starts up. But I'll pull that stuff out of this PR.

**Review by tonybaloney — APPROVED** [review]

**Pull request merged**